### PR TITLE
Implement Web Login for SSO

### PR DESCRIPTION
**Pull request by shardulm94**
A shot at implementing SSO support which should help towards #4, #9, #18 and maybe #14. I don't have a username and password for safari to check if I broke something in the username and pass auth codepath.

### CODE REVIEW

**Comment by geahaad** [comment]
Awesome, @shardulm94, works like a charm with my company's SSO.
+1 to merge this request! ✌️

**Comment by nicohaenggi** [comment]
@shardulm94 Thanks alot for your effort. I will go through it and merge it later this week!

**Comment by TehPsychedelic** [comment]
Any word on when this will be merged into master?

**Comment by xevenheaven** [comment]
Any updates for SSO support? Unfortunately this PR didn't work for me. 😔 

Getting this error:
```
XMLHttpRequest cannot load https://www.safaribooksonline.com/api/v1. Redirect from 'https://www.safaribooksonline.com/api/v1' to 'https://www.safaribooksonline.com/api/v1/' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.
```
Console logs just show:
```
[safari-downloader] Checking if authenticated...
[safari-downloader] Not authenticated yet
[safari-downloader] Checking if authenticated...
[safari-downloader] Not authenticated yet
[safari-downloader] Checking if authenticated...
[safari-downloader] Error while checking
[safari-downloader] Checking if authenticated...
[safari-downloader] Error while checking
```

**Comment by ohthehugemanatee** [comment]
Works great for me with a microsoft.com account. I second the vote to merge!

I had a separate problem after authentication, but it did download part of the book on the first try. And it successfully used the cached cookies on subsequent tries.

**Comment by invious** [comment]
```
XMLHttpRequest cannot load https://www.safaribooksonline.com/api/v1. Redirect from 'https://www.safaribooksonline.com/api/v1' to 'https://www.safaribooksonline.com/api/v1/' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin 'https://learning.oreilly.com' is therefore not allowed access.
```

**Comment by njerig** [comment]
Can someone explain how this resolves #14? I eventually get this error
```
[safari-downloader] StatusCodeError: 401 - {"detail":"Authentication credentials were not provided."}
```
after my cookies are loaded into `.cookies-file`. (I'm trying to download books from proquest.safaribooks.com.)

File: lib/safari-web/index.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+  })
+}
+
+SafariWeb.prototype.login = function login() {
+  return new Promise(function(resolve) {
+    this.nightmare
+    .on('console', (log, msg) => {
+      logger.log(msg)
+    })
+    .goto('https://safaribooksonline.com')
```
Comment (rcaval) [inline lib/safari-web/index.js]: This needs to change to `https://learning.oreilly.com` for it to work

File: lib/safari-web/index.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+            } else if (xmlhttp.status == 401) {
+              console.log("Not authenticated yet")
+              resolveWait(false)
+            } else {
+              console.log("Error while checking")
+              resolveWait(false)
+            }
+          }
+        };
+        xmlhttp.open("GET", "https://www.safaribooksonline.com/api/v1", true);
```
Comment (rcaval) [inline lib/safari-web/index.js]: This here also needs to change:
```suggestion
        xmlhttp.open("GET", "https://learning.oreilly.com/api/v1", true);
```

File: lib/safari/index.js [hunk 3]
```diff
@@ -113,13 +120,13 @@ Safari.prototype.authorizeUser = function authorizeUser(username, password) {
 */
 Safari.prototype.fetchResource = function fetchResource(url, options) {
   debug(`fetchResource called with URL: ${url}`);
-  if(!url || !this.accessToken) return Promise.reject("url was not specified or user has not been authorized yet");
+  if(!url || !this.accessToken && !this.cookieString) return Promise.reject("url was not specified or user has not been authorized yet");
   // ## prepare options for resource request
   var uri = `${this.baseUrl}/${url}`;
```
Comment (rcaval) [inline lib/safari/index.js]: For some reason, either the request lib or the api cannot handle double slashes
```suggestion
  var uri = `${this.baseUrl}${url}`;
```

**Comment by JarnoRFB** [comment]
Would it be possible to also enable this for `ACADEMIC LICENSE &
PUBLIC LIBRARY USERS` which is an option on this site https://proquestcombo.safaribooksonline.com/home ?

**Comment by yeikel** [comment]
It does not seem to work for me 

[safari-downloader] [object Object]
[safari-downloader] [Facebook Pixel] - Duplicate Pixel ID: xx.
[safari-downloader] %cElectron Security Warning (Insecure Content-Security-Policy)
[safari-downloader] Checking if authenticated...
[safari-downloader] Error while checking
[safari-downloader] Checking if authenticated...
[safari-downloader] Error while checking
[safari-downloader] Checking if authenticated...
[safari-downloader] Error while checking
[safari-downloader] Checking if authenticated...
[safari-downloader] Error while checking
[safari-downloader] [object Object]
[safari-downloader] [Facebook Pixel] - Duplicate Pixel ID: xxxx.
[safari-downloader] %cElectron Security Warning (Insecure Content-Security-Policy)
[safari-downloader] Checking if authenticated...
[safari-downloader] Error while checking
[safari-downloader] [Facebook Pixel] - Duplicate Pixel ID: xxx.
[safari-downloader] %cElectron Security Warning (Insecure Content-Security-Policy)
[safari-downloader] [object Object]
[safari-downloader] Checking if authenticated...
[safari-downloader] Error while checking
[safari-downloader] Checking if authenticated...
[safari-downloader] Error while checking

File: lib/safari-web/index.js (hunk 1 above)
Comment (yeikel) [reply lib/safari-web/index.js]: Even with this change it did not work for me

**Pull request closed without merging**